framework/include/linesearches/ContactLineSearch.h

```diff
@@ -0,0 +1,85 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef CONTACTLINESEARCH_H
+#define CONTACTLINESEARCH_H
+
+#include "LineSearch.h"
+
+/**
+ * This class implements a custom line search for use with
+ * mechanical contact. The line search is not fancy. It takes two parameters, set in the MOOSE
+ * Executioner block: `contact_line_search_ltol` and `contact_line_search_allowed_lambda_cuts`. The
+ * allowed_lambda_cuts parameter specifies the number of times the line search is allowed to cut
+ * lambda. If allowed to be cut, lambda will be reduced by half, and a new residual will be
+ * evaluated. If the residual is smaller with a smaller lambda, then cuts will continue until
+ * reaching allowed_lambda_cuts. If the residual is larger with a smaller lambda, then the line
+ * search is curtailed and the smaller residual is used. It's recommended that allowed_lambda_cuts
+ * be <= 3, with smaller values being used for smaller contact problems. This is to allow necessary
+ * residual increases when the transient problem requires significant changes in the contact state.
+ *
+ * When the contact set is changing, the user may optionally use a looser linear tolerance set by
+ * the `contact_line_search_ltol` parameter. Then when the contact set is changing during the
+ * beginning of the Newton solve, unnecessary computational expense is avoided. Then when the
+ * contact set is resolved late in the Newton solve, the linear tolerance will return to the finer
+ * tolerance set through the traditional `l_tol` parameter.
+ */
+class ContactLineSearch : public LineSearch
+{
+public:
+  ContactLineSearch(FEProblemBase & fe_problem,
+                    MooseApp & app,
+                    size_t allowed_lambda_cuts,
+                    Real contact_ltol,
+                    bool affect_ltol);
+
+  // A dummy pure virtual destructor because this must be a base class; the derived class must
+  // implement the line searching capabilities
+  virtual ~ContactLineSearch() = 0;
+
+  /**
+   * The current contact state
+   */
+  std::shared_ptr<std::set<dof_id_type>> contact_state() { return _current_contact_state; }
+
+  /**
+   * The old contact state
+   */
+  std::set<dof_id_type> & old_contact_state() { return _old_contact_state; }
+
+  /**
+   * Method for printing the contact information
+   */
+  void printContactInfo();
+
+protected:
+  /// The current contact set
+  std::shared_ptr<std::set<dof_id_type>> _current_contact_state;
+  /// The old contact set
+  std::set<dof_id_type> _old_contact_state;
```

Comment:
This is bizarre, one shared pointer, and one instance variable to hold the same kind of data. BTW - it looks like *ownership* of the set is never shared. Yes the data is shared, but not the ownership of the set itself. You should probably just let these both be instance variables and return a reference (or maybe a raw pointer) when the underlying data needs to be shared.